torch/lib/c10d/reducer.cpp

```diff
@@ -969,6 +1000,24 @@ void Reducer::populate_bucket_views_out(
   }
 }
 
+void Reducer::prepare_for_forward() {
+  std::lock_guard<std::mutex> lock(mutex_);
+  num_iterations_++;
+  if (replicas_[0][0].is_cuda()) {
+    #ifdef USE_CUDA
+    // Record event only for single process single device
+    // and single device module.
+    if (replicas_.size() == 1 && !is_multi_device_module_) {
+      // Create and record event on the replicas_[0][0].device().
+      at::cuda::set_device(replicas_[0][0].device().index());
```

Comment:
Nit: I believe the general pattern is to use a `DeviceGuard`.

Reply:
Actually this is the root cause where deadlock happens for single process single device. when users are not explicitly set service in their python codes, the current device index will always be 0 for all ranks. In this case, rank 1 will try to create event on device 0, but rank 0 has launched nccl all reduce on device 0 ==> this will cause deadlock. 

DeviceGuard will not work here as well. It may be one of two reasons: 1) cudaEvent record internally used cudaGuard, which may be conflict with DeviceGuard outside. 2) DeviceGuard constructor will reset device to active device during DeviceGuard construction, which may cause issue (I am not sure). 

So I gave up using DeviceGuard. On the other hand, I think set_device() call should be safe here, as we only set_device() for single process single device program that usually requires users to explicitly set device on python upfront as well.

Reply:
Nice catch! Sounds like the deadlock issue is quite similar to what we discussed offline with the deadlock in the profiler as well.